File: clients/typescript/src/cache_manager.ts
import { IConnectionManager } from './topic_manager';

export interface ICacheWatcher {
    readonly group: string;
    dispatchValueChanged(key: string, value: any): void;
}

export class CacheManager {
    private readonly connectionManager: IConnectionManager;
    public readonly cacheMap: Map<string, Set<ICacheWatcher>> = new Map();

    constructor(connectionManager: IConnectionManager) {
        this.connectionManager = connectionManager;
    }

    public subscribeWatcher(watcherInstance: ICacheWatcher): void {
        const groupName = watcherInstance.group;
        if (!groupName)
            return;

        if (!this.cacheMap.has(groupName))
            this.cacheMap.set(groupName, new Set());

        const watcherSet = this.cacheMap.get(groupName)!;
        const isCacheEmpty = watcherSet.size === 0;
        watcherSet.add(watcherInstance);

        if (isCacheEmpty && this.connectionManager.isConnected)
            this.connectionManager.send('Cache.Subscribe', groupName);
    }

    public unsubscribeWatcher(watcherInstance: ICacheWatcher): void {
        const groupName = watcherInstance.group;
        if (!groupName || !this.cacheMap.has(groupName))
            return;

        const watcherSet = this.cacheMap.get(groupName)!;
        watcherSet.delete(watcherInstance);

        if (watcherSet.size === 0) {
            this.cacheMap.delete(groupName);
            if (this.connectionManager.isConnected)
                this.connectionManager.send('Cache.Unsubscribe', groupName);
        }
    }

    public resubscribeAll(): void {
        if (!this.connectionManager.isConnected)
            return;

        for (const [groupName, watcherSet] of this.cacheMap.entries()) {
            if (watcherSet.size > 0)
                this.connectionManager.send('Cache.Subscribe', groupName);
        }
    }

    public dispatchValueChanged(groupName: string, key: string, value: any): void {
        const watcherSet = this.cacheMap.get(groupName);
        if (!watcherSet)
            return;

        for (const watcherInstance of Array.from(watcherSet))
            watcherInstance.dispatchValueChanged(key, value);
    }
}

File: clients/csharp/AmenoLink/CacheManager.cs
namespace AmenoLink;

internal class CacheManager(IConnectionManager connectionManager)
{
    private readonly Dictionary<string, HashSet<ICacheWatcherSubscriber>> cacheMap = [];
    private readonly Lock lockObject = new();

    public void SubscribeWatcher(ICacheWatcherSubscriber watcherInstance)
    {
        string groupName = watcherInstance.Group;
        if (string.IsNullOrEmpty(groupName))
            return;

        bool isCacheEmpty = false;
        lock (lockObject)
        {
            if (!cacheMap.TryGetValue(groupName, out var watcherSet))
            {
                watcherSet = [];
                cacheMap[groupName] = watcherSet;
            }

            isCacheEmpty = watcherSet.Count == 0;
            watcherSet.Add(watcherInstance);
        }

        if (isCacheEmpty && connectionManager.IsConnected)
            _ = connectionManager.Send("Cache.Subscribe", groupName);
    }

    public void UnsubscribeWatcher(ICacheWatcherSubscriber watcherInstance)
    {
        string groupName = watcherInstance.Group;
        if (string.IsNullOrEmpty(groupName))
            return;

        bool shouldUnsubscribeOnServer = false;
        lock (lockObject)
        {
            if (!cacheMap.TryGetValue(groupName, out var watcherSet))
                return;

            watcherSet.Remove(watcherInstance);

            if (watcherSet.Count == 0)
            {
                cacheMap.Remove(groupName);
                shouldUnsubscribeOnServer = true;
            }
        }

        if (shouldUnsubscribeOnServer && connectionManager.IsConnected)
            _ = connectionManager.Send("Cache.Unsubscribe", groupName);
    }

    public void ResubscribeAll()
    {
        if (!connectionManager.IsConnected)
            return;

        List<string> groupsToResubscribe;
        lock (lockObject)
        {
            groupsToResubscribe = cacheMap.Where(kvp => kvp.Value.Count > 0).Select(kvp => kvp.Key).ToList();
        }

        foreach (string groupName in groupsToResubscribe)
            _ = connectionManager.Send("Cache.Subscribe", groupName);
    }

    public void DispatchValueChanged(string groupName, string key, object? value)
    {
        List<ICacheWatcherSubscriber> targets;
        lock (lockObject)
        {
            if (!cacheMap.TryGetValue(groupName, out var watcherSet))
                return;

            targets = [.. watcherSet];
        }

        foreach (var watcherInstance in targets)
            watcherInstance.DispatchValueChanged(key, value);
    }
}
